docs/public-api.md

```diff
@@ -0,0 +1,65 @@
+
+  What's the point of this restriction? If we prefix something with an underscore or put it into one of our internal modules, we can't commit to an API for it and don't expect users to ever use this. Also, there might be different underscored functions/types depending on the platform.
+  
+  Examples:
+   - ✅ `channel.close(promise: nil)`
+   - ❌ `channel._unsafe.flush0()`, underscored property
+   - ❌ `import CNIOAtomics`, module name doesn't start with NIO
+- it is acceptable to conform NIO types to protocols _you control_, ie. a protocol in _your codebase_. It's not acceptable to confirm any NIO types to protocols in the standard library or outside of your control.
+
+  What's the point of this restriction? NIO in a later new minor version might conform its type to said protocol.
+  
+  Examples:
+    - ✅ `extension EventLoopFuture: MyOwnProtocol { ... }`, assuming `MyOwnProtocol` lives in your codebase
+    - ❌ `extension EventLoopFuture: DebugStringConvertible { ... }`, `DebugStringConvertible` is a standard library protocol
+- it is acceptable to conform _your own types_ to NIO protocols but it's not acceptable to conform types you do not control to NIO protocols.
+
+  What's the point of this restriction? Again, NIO might later add this conformance.
+
+  Examples:
+    - ✅ `extension MyHandler: ChannelHandler { ... }`
+    - ❌ `extension Array: EventLoopGroup where Element: EventLoop { ... }`, as `Array` lives in the standard library
+
+- it is acceptable to add extensions to NIO types that either use one of your types as a non-default argument or are prefixed in a way that prevents clashes
+
+  What's the point of these restrictions? NIO might later add a member function which might then clash with yours and might stop your code from compiling and/or working.
+
+  Examples:
+    - ✅ `extension ByteBuffer { mutating func readMyType(at: Int) -> MyType }`, acceptable because it returns a `MyType`
+    - ✅ `extension ByteBuffer { mutating func myProjectReadInteger(at: Int) -> Int`, acceptable because prefixed with `myProject`
+    - ❌ `extension ByteBuffer { mutating func readBool(at: Int) -> Bool }`, because `Bool` is a standard library type not in your control
+
+### Promises the NIO Team Make
+
+Before releasing a new major version, ie. SwiftNIO 3.0.0 we promise the following:
+
+- no global types/modules/functions will be added that do not have `NIO*`/`nio*`/`CNIO*`/`cnio*`/`c_nio*` prefixes, this includes C functions that are not in the public API.
+  Examples:
+    - we might add a new global type called `NIOJetPack`
+    - we will not add a global new type called `Tomato`
+    - we might add a new module called `NIOHelpers`
+    - we will not add a new module called `Helpers`
+    - we might add a new global functions called `nioRun` (very unlikely) or `c_nio_SSL_write` (more likely)
+    - we will not add a new global function called `SSL_write`
```

Comment:
Hm. But what if SwiftNIO grows a new exported dependency?

Reply:
if we vendor new stuff, we symbol-prefix everything:
- https://github.com/apple/swift-nio/blob/master/Sources/CNIOHTTPParser/update_and_patch_http_parser.sh#L48-L65
- https://github.com/apple/swift-nio/blob/master/Sources/CNIOSHA1/update_and_patch_sha1.sh#L43
- https://forums.swift.org/t/rfc-moving-swiftnio-ssl-to-boringssl/18280

Reply:
What if your *dependencies* grow a new exported dependency?

Reply:
@jrose-apple thank you, this is very valuable! So what I'm trying to say (and I'll try to rephrase this) is that we won't add any _C symbols_ into our binaries. We have done this (`http_parser.c` wasn't name-spaced in NIO 1.0.0) and it lead to problems because people combined NIO with something else that also vendored a non-name-spaced `http_parser.c`.
I understand that Swift's namespacing model doesn't allow those guarantees but so far Swift mangles enough information into the symbol names that we didn't suffer any of those clashes in reality. For the vendored C stuff however this was totally different and our users have actually hit this which is why we symbol-prefix everything C now.

We're also in a pretty unique spot with NIO that we don't have many dependencies at all. NIO core dynamically links the Swift stdlib, runtime, etc (not Foundation), zlib, libc and nothing else and we're not changing this. The rule has always been: Anything that depends on other system libraries goes into a separate package (therefore `swift-nio-ssl` and `swift-nio-transport-services`). And I don't think swift-nio-ssl/transport-services should grow a new dependency without a fresh major version. All of our packages so far only depend on only NIO stuff.

I believe this should basically get us to a place where if NIO 2.x.m works fine, the user types `swift package update` gets 2.y.n (where y >= x) and there's no way it doesn't work (unless the user has their own modules/C functions with `CNIO`,`c_nio`,`NIO`, ... prefixes and we added one that clashed). Do you think this is accurate?

```
$ nm ./.build/x86_64-apple-macosx10.10/debug/NIOTLSServer | grep ' T ' | grep -v -e NIO -e catmc -e c_nio
0000000100000000 T __mh_execute_header
00000001001a8010 T _main
$ nm ./.build/x86_64-apple-macosx10.10/debug/NIOHTTP1Server | grep ' T ' | grep -v -e NIO -e catmc -e c_nio
0000000100000000 T __mh_execute_header
00000001001aecc0 T _main
```

(`catmc` is for historical reasons the prefix for our C atomics but they have been there since 1.0.0 so nothing got added and we should really rename them to `c_nio_`)

Reply:
I guess it's just as possible that zlib adds an unprefixed function, but because you control your dependencies so tightly you might be able to get away with this.